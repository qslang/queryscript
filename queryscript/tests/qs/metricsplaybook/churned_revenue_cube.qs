import * from data;
import metrics_cube from cube;

fn activity_filter(activity_type string) {
      select
        c.id as customer_id,
        c.segment,
        c.channel,
        c.cohort,
        m.timestamp,
        m.revenue_impact,
        m.activity,
        m.plan_type
    from
        "contract_stream" m
        join "dim_customer" c
            on m.customer_id = c.id
    where
        m.activity = activity_type
}

let churned_revenue_cube = metrics_cube(
    ['month', 'day'],
    ['segment', 'channel', 'plan_type'],
    activity_filter('customer_churn_committed'),
    sum,
    'revenue_impact',
);

let new_revenue_cube = metrics_cube(
    ['month', 'day'],
    ['segment', 'channel', 'plan_type'],
    activity_filter('new_contract_started'),
    sum,
    'revenue_impact',
);

let contraction_revenue_cube = metrics_cube(
    ['month', 'day'],
    ['segment', 'channel', 'plan_type'],
    activity_filter('contraction_contract_started'),
    sum,
    'revenue_impact',
);


let expansion_contract_started = metrics_cube(
    ['month', 'day'],
    ['segment', 'channel', 'plan_type'],
    activity_filter('expansion_contract_started'),
    sum,
    'revenue_impact',
);

SELECT COUNT(*) FROM churned_revenue_cube;
SELECT COUNT(*) FROM new_revenue_cube;
SELECT COUNT(*) FROM contraction_revenue_cube;
SELECT COUNT(*) FROM expansion_contract_started;


let churned_bar_graph = churned_revenue_cube WITH VIZ {
    mark: 'bar',
    encoding: {
        x: { field: 'metric_date' , type: 'nominal'    },
        y: { field: 'metric_value', type: 'quantitative' },
    },
};
