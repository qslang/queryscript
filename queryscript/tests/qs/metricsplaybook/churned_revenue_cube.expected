{
    "compile_errors": [],
    "decls": {
        "let activity_filter": λ {
        	activity_type Utf8,
        } -> [{
        	customer_id Int64,
        	segment Utf8,
        	channel Utf8,
        	cohort Utf8,
        	timestamp Timestamp(
            Microsecond,
            None,
        ),
        	revenue_impact Float64,
        	activity Utf8,
        	plan_type Utf8,
        }],
        "let churned_revenue_cube": [{
        	metric_model Utf8,
        	is_snapshot_reliant_metric Boolean,
        	anchor_date Utf8,
        	date_grain Utf8,
        	metric_date Date32,
        	slice_object Utf8,
        	slice_dimension Utf8,
        	slice_value Utf8,
        	metric_calculation Utf8,
        	metric_value Float64,
        }],
        "let contract_stream": [{
        	id Int64,
        	customer_id Int64,
        	timestamp Timestamp(
            Microsecond,
            None,
        ),
        	activity Utf8,
        	revenue_impact Float64,
        	plan_type Utf8,
        	activity_occurrence Int64,
        	activity_repeated_at Timestamp(
            Microsecond,
            None,
        ),
        }],
        "let contraction_revenue_cube": [{
        	metric_model Utf8,
        	is_snapshot_reliant_metric Boolean,
        	anchor_date Utf8,
        	date_grain Utf8,
        	metric_date Date32,
        	slice_object Utf8,
        	slice_dimension Utf8,
        	slice_value Utf8,
        	metric_calculation Utf8,
        	metric_value Float64,
        }],
        "let dim_customer": [{
        	id Int64,
        	segment Utf8,
        	channel Utf8,
        	first_contract_signed_date Timestamp(
            Microsecond,
            None,
        ),
        	cohort Utf8,
        }],
        "let expansion_contract_started": [{
        	metric_model Utf8,
        	is_snapshot_reliant_metric Boolean,
        	anchor_date Utf8,
        	date_grain Utf8,
        	metric_date Date32,
        	slice_object Utf8,
        	slice_dimension Utf8,
        	slice_value Utf8,
        	metric_calculation Utf8,
        	metric_value Float64,
        }],
        "let metrics_cube": ∀ "__Return", "R", "S" λ {
        	date_slices [Utf8],
        	metric_slices [Utf8],
        	source "R",
        	metric_fn "S",
        	metric_name Utf8,
        } -> "__Return",
        "let new_revenue_cube": [{
        	metric_model Utf8,
        	is_snapshot_reliant_metric Boolean,
        	anchor_date Utf8,
        	date_grain Utf8,
        	metric_date Date32,
        	slice_object Utf8,
        	slice_dimension Utf8,
        	slice_value Utf8,
        	metric_calculation Utf8,
        	metric_value Float64,
        }],
        "let src_contract_stream": [{
        	id Int64,
        	customer_id Int64,
        	timestamp Timestamp(
            Microsecond,
            None,
        ),
        	activity Utf8,
        	revenue_impact Float64,
        	plan_type Utf8,
        	activity_occurrence Int64,
        	activity_repeated_at Timestamp(
            Microsecond,
            None,
        ),
        }],
        "let src_dim_customer": [{
        	id Int64,
        	segment Utf8,
        	channel Utf8,
        	first_contract_signed_date Timestamp(
            Microsecond,
            None,
        ),
        }],
        "type ContractStream": {
        	id Int64,
        	customer_id Int64,
        	timestamp Timestamp(
            Microsecond,
            None,
        ),
        	activity Utf8,
        	revenue_impact Float64,
        	plan_type Utf8,
        	activity_occurrence Int64,
        	activity_repeated_at Timestamp(
            Microsecond,
            None,
        ),
        },
        "type DimCustomer": {
        	id Int64,
        	segment Utf8,
        	channel Utf8,
        	first_contract_signed_date Timestamp(
            Microsecond,
            None,
        ),
        },
    },
    "queries": [
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "COUNT(*)",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| COUNT(*) |\n|----------|\n| 272      |",
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "COUNT(*)",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| COUNT(*) |\n|----------|\n| 276      |",
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "COUNT(*)",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| COUNT(*) |\n|----------|\n| 249      |",
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "COUNT(*)",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| COUNT(*) |\n|----------|\n| 257      |",
            },
        ),
    ],
}